src/algebraic_geometry/EllipticCurve/point.lean

```diff
@@ -0,0 +1,417 @@
+group operations by explicit equations and proves that the set is an abelian group (TODO).
+
+## Main definitions
+
+ * `EllipticCurve.point`: a `K`-rational point on `E`.
+ * `EllipticCurve.point.add`: addition of two `K`-rational points on `E`.
+
+## Main statements
+
+ * `EllipticCurve.point.add_comm`: addition of two `K`-rational points on `E` is commutative.
+ * TODO: addition of two `K`-rational points on `E` is associative (HARD), and hence forms a group.
+
+## Notations
+
+ * `E⟮K⟯`: the group of `K`-rational points on an elliptic curve `E`.
+
+## Implementation notes
+
+The proofs of `EllipticCurve.point.weierstrass_dbl'`, `EllipticCurve.point.weierstrass_add'`, and
+`EllipticCurve.point.add_assoc` (TODO) are slow due to the sheer size of the polynomials involved,
+but they can be hastened considerably by breaking the computations into multiple steps if necessary.
+
+## References
+
+[J Silverman, *The Arithmetic of Elliptic Curves*][silverman2009]
+
+## Tags
+
+elliptic curve, rational point, group law
+-/
+
+universes u v
+
+namespace EllipticCurve
+
+section basic
+
+/-!
+### Points in `E⟮K⟯`
+
+Use `0` instead of `EllipticCurve.point.zero`.
+-/
+
+open polynomial
+
+open_locale polynomial
+
+variables {F : Type u} [comm_ring F] (E : EllipticCurve F) {K : Type v} [comm_ring K] [algebra F K]
+
+/-- The Weierstrass polynomial $Y^2 + a_1XY + a_3Y - (X^3 + a_2X^2 + a_4X + a_6)$ of `E`. -/
+noncomputable def weierstrass_polynomial : K[X][X] :=
+X ^ 2 + C (C ↑E.a₁ * X) * X + C (C ↑E.a₃) * X - C (X ^ 3 + C ↑E.a₂ * X ^ 2 + C ↑E.a₄ * X + C ↑E.a₆)
+
+/-- The proposition that an affine point $(x, y)$ lies in `E`.
+In other words, the Weierstrass polynomial of `E` evaluates to zero at $(x, y)$. -/
+def weierstrass_equation (x y : K) : Prop := eval x (eval (C y) E.weierstrass_polynomial) = 0
+
+@[simp] lemma eval_weierstrass_polynomial (x y : K) :
+  eval x (eval (C y) E.weierstrass_polynomial)
+    = y ^ 2 + ↑E.a₁ * x * y + ↑E.a₃ * y - (x ^ 3 + ↑E.a₂ * x ^ 2 + ↑E.a₄ * x + ↑E.a₆) :=
+by simp only [weierstrass_polynomial, eval_sub, eval_add, eval_mul, eval_pow, eval_C, eval_X]
+
+lemma weierstrass_equation_iff' (x y : K) :
+  E.weierstrass_equation x y
+    ↔ y ^ 2 + ↑E.a₁ * x * y + ↑E.a₃ * y - (x ^ 3 + ↑E.a₂ * x ^ 2 + ↑E.a₄ * x + ↑E.a₆) = 0 :=
+by rw [weierstrass_equation, eval_weierstrass_polynomial]
+
+@[simp] lemma weierstrass_equation_iff (x y : K) :
+  E.weierstrass_equation x y
+    ↔ y ^ 2 + ↑E.a₁ * x * y + ↑E.a₃ * y = x ^ 3 + ↑E.a₂ * x ^ 2 + ↑E.a₄ * x + ↑E.a₆ :=
+by rw [weierstrass_equation_iff', sub_eq_zero]
+
+variables (K)
+
+/-- The group of `K`-rational points `E⟮K⟯` on an elliptic curve `E` over `F`. This consists of the
+unique point at infinity `EllipticCurve.point.zero`and the affine points `EllipticCurve.point.some`
+satisfying the Weierstrass equation $y^2 + a_1xy + a_3y = x^3 + a_2x^2 + a_4x + a_6$. -/
+inductive point
+| zero
+| some (x y : K) (h : E.weierstrass_equation x y)
```

Comment:
I thought of this, but the pattern match in `add` becomes awkward `if hx : x1 = x2 ...` - any idea how to go around this?

Reply:
You would have to write `| (@some _ _ _ _ _ _ x₁ y₁ h₁) (@some _ _ _ _ _ _ x₂ y₂ h₂) :=`, admittedly quite lengthy, but this is the only place you need to do so. If you only use one ring rather than two, there will be fewer underscores.

Reply:
Sounds good. It doesn't seem like I need to do `rcases` in the two proofs though.